File: CeliTax/TaxYear.m
//
//  TaxYear.m
//  CeliTax
//
//  Created by Leon Chen on 2015-07-15.
//  Copyright (c) 2015 CraveNSave. All rights reserved.
//

#import "TaxYear.h"

#define kKeyTaxYear         @"TaxYear"
#define kKeyDataAction      @"DataAction"

@implementation TaxYear

- (void) encodeWithCoder: (NSCoder *) coder
{
    [coder encodeInteger: self.taxYear forKey: kKeyTaxYear];
    [coder encodeInteger:self.dataAction forKey:kKeyDataAction];
}

- (instancetype) initWithCoder: (NSCoder *) coder
{
    if (self = [self init])
    {
        self.taxYear = [coder decodeIntegerForKey: kKeyTaxYear];
        
        self.dataAction = [coder decodeIntegerForKey:kKeyDataAction];
    }
    
    return self;
}

- (id) copyWithZone: (NSZone *) zone
{
    TaxYear *copy = [[[self class] alloc] init];
    
    if (copy)
    {
        copy.taxYear = self.taxYear;
        copy.dataAction = self.dataAction;
    }
    
    return copy;
}

- (NSDictionary *) toJson
{
    NSMutableDictionary *json = [NSMutableDictionary dictionary];
    
    json[kKeyTaxYear] = @(self.taxYear);
    
    json[kKeyDataAction] = @(self.dataAction);
    
    return json;
}

@end

File: CeliTax/TaxYear.swift
//
//  TaxYear.swift
//  CeliTax
//
//  Created by Leon Chen on 2015-10-31.
//  Copyright © 2015 CraveNSave. All rights reserved.
//

import Foundation

@objc
class TaxYear : NSObject, NSCoding, NSCopying //TODO: Remove Subclass to NSObject when the entire app has been converted to Swift
{
    let kKeyTaxYear : String = "TaxYear"
    let kKeyDataAction : String = "DataAction"
    
    var taxYear : Int = 0
    var dataAction : DataActionStatus = DataActionStatus.DataActionNone
    
    override init()
    {
        
    }
    
    required init(coder decoder: NSCoder)
    {
        self.taxYear = decoder.decodeIntegerForKey(kKeyTaxYear)
        self.dataAction = DataActionStatus(rawValue: decoder.decodeIntegerForKey(kKeyDataAction))!
    }
    
    func encodeWithCoder(coder: NSCoder)
    {
        coder.encodeInteger(self.taxYear, forKey: kKeyTaxYear)
        coder.encodeInteger(self.dataAction.rawValue, forKey: kKeyDataAction)
    }
    
    func copyWithZone(zone: NSZone) -> AnyObject
    {
        let copy = TaxYear()
        
        copy.taxYear = self.taxYear
        copy.dataAction = self.dataAction
        
        return copy
    }
    
    func toJson() -> Dictionary<String, AnyObject>
    {
        var json = Dictionary<String, AnyObject>()
        
        json[kKeyTaxYear] = self.taxYear
        json[kKeyDataAction] = self.dataAction.rawValue
        
        return json
    }
}